### PR TITLE
Refactor: Kyber internals

### PR DESCRIPTION
**Pull request by reneme**
This is an iteration of the Kyber internals, in preparation of an ML-KEM implementation. Also, it retrofits the helpers we added along the way in the last two years. Most notably: `BufferStuffer` and `BufferSlicer` and `Strong<>` types. Especially the strong types made it possible to restrict the "symmetric algorithms" in `kyber_symmetric_primitives.h`.

Before, most of the algorithm (i.e. the Polynomial structures) was hidden away in the `kyber.cpp` compilation unit in the `kyber_common` module. For added flexibility, I restructured the code into a few distinct header and implementation files. Also, the `kyber` and `kyber_90s` modules now moved inside a new internal `kyber_round3` module. These changes make room for an upcoming ML-KEM-ipd implementation that reuses as much of the implementation as possible.

The public interface isn't affected by any of those changes.

### CODE REVIEW

**Comment by reneme** [comment]
Force-pushed after #3888, to fix CI.

**Review by FAlbertDev — APPROVED** [review]
I really like Kyber with strong types! Also, the buffer stuffer fits very well. No more nasty pointer arithmetic 🎉

File: src/lib/pubkey/kyber/kyber_common/kyber.cpp [hunk 1]
```diff
@@ -597,14 +642,37 @@ class PolynomialVector {
...
+         for(size_t i = 0; i < length; ++i) {
+            fn(vecs.m_vec[i]...);
+         }
+      }
+
+      template <typename FnT, typename... OtherPolynomialVectorTs>
+         requires all_same_v<PolynomialVector, std::remove_cvref_t<OtherPolynomialVectorTs>...>
+      void on_all(FnT&& fn, OtherPolynomialVectorTs&&... other_vecs) {
+         for_all(fn, *this, std::forward<OtherPolynomialVectorTs>(other_vecs)...);
+      }
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber.cpp:663]: Mhm, I am on the fence with that one as well. I spiked it, saw that it works and then left it in.

In my opinion it does improve readability at the using site somewhat; it clarifies on first glance that most methods are really just delegates to the individual `Polynomial` objects. But, certainly, the implementation of the `on_all()` method does not really justify that (admittedly very minor) gain.

Will remove it again.

File: src/lib/pubkey/kyber/kyber_common/kyber.cpp [hunk 2]
```diff
@@ -616,44 +684,39 @@ class PolynomialVector {
...
+               PolynomialVector r(mode.k());
+               r.on_all([&](auto& p) { p = Polynomial::getnoise_eta1(s, nonce++, mode); });
+               return r;
+            },
+            seed);
       }
 
-      template <typename T = std::vector<uint8_t>>
+      template <concepts::resizable_byte_buffer T = std::vector<uint8_t>>
       T to_bytes() {
```
[the remark this answers is not in the document]
Comment (reneme) [reply src/lib/pubkey/kyber/kyber_common/kyber.cpp:714]: Fair, though. Thanks.

**Review by FAlbertDev — APPROVED** [review]
Nice! I have the code on my left and the FIPS 203 spec on my right screen, and can easily compare it 👍

**Comment by reneme** [comment]
Rebased and resolved conflicts after #3874 got merged.

**Review by randombit — APPROVED** [review]
Very nice :+1:

**Comment by reneme** [comment]
Thanks for the review! We have a further iteration for this in the pipeline, though.

It'll be a few weeks before we're done with it, unfortunately. Currently busy preparing the concluding workshop for BSI's project 481.

**Comment by reneme** [comment]
Resolved a conflict with master.

**Comment by reneme** [comment]
I realized that the planned changes, I mentioned last month, can be easily added on top of this. So, let's merge this now and we'll continue working on master.

**Pull request merged**